include/seqan3/contrib/parallel/latch.hpp

```diff
@@ -0,0 +1,164 @@
+ * constructed. An individual latch is a single-use object; once the count has been reached, the latch cannot be reused.
+ * This implementation uses a lock-free mechanism if the atomic operations on the respective platform are lock-free.
+ *
+ * \note The adapts the proposal for the c++ standard
+ * [P0666R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0666r2.pdf) for latches.
+ */
+class latch
+{
+public:
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    latch()                           = delete;  //!> Deleted.
+    latch(latch const &)              = delete;  //!> Deleted.
+    latch(latch &&)                   = default; //!> Defaulted.
+    latch & operator=(latch const &)  = delete;  //!> Deleted.
+    latch & operator=(latch &&)       = default; //!> Defaulted.
+
+    //!\brief Destructs the latch and waits for all participating threads to arrive.
+    ~latch()
+    {
+        spin_delay delay{};
+        while (num_waiting.load(std::memory_order_acquire) > 0)
+            delay.wait();
+    }
+
+    /*!\brief Constructs the latch with the expected number of threads.
+     * \param expected The number of threads participating in this synchronisation point.
+     */
+    explicit latch(ptrdiff_t const expected) :
+        counter{expected}
+    {
+        assert(expected >= 0);
+        num_waiting.store(0, std::memory_order_relaxed);
+    }
+    //!\}
+
+    /*!\brief Atomically decrements counter by `n`.
+     * \param n The value to subtract.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point with `n` count without waiting. If `n` is less than 0 or greater then the
+     * current count, then this operation results in undefined behaviour for all other participating threads.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void arrive(ptrdiff_t n = 1) noexcept
+    {
+        assert(counter.load(std::memory_order_acquire) >= n);
+        assert(counter.load(std::memory_order_acquire) >= 0);
+
+        counter.fetch_sub(n, std::memory_order_acq_rel);
+    }
+
+    /*!\brief Atomically decrements counter by `n` and blocks the calling thread.
+     * \param n The value to subtract.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point with `n` count and blocks the calling thread until all participating threads
+     * have arrived. If `n` is less than 0 or greater then the current count, then this operation results in undefined
+     * behaviour for all other participating threads.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void arrive_and_wait(ptrdiff_t n = 1) noexcept
+    {
+        ++num_waiting; // ensure that destructor is not finished in-between the arrive and wait call.
```

Comment:
> // ensure that destructor is not finished in-between the arrive and wait call.

Is this a todo or a gurantee?

Reply:
Well, it is a little confusing. The proposal says that the destructor should wait for all threads to leave wait. Of course this can't be guaranteed, since a thread could be stalled at any time by  a preemptive thread scheduler, thus after decrementing num_waiting the thread could lose control and the destructor could finish it's job. What it actually tries to prevent is that a thread is still waiting in the loop while the destructor is being called. I think there is no solution to that except to protect it from outside, but I took a glance into the reference implementation and they did like so. So I thought I take this as well.